File: TestUtils/src/main/java/com/braintreepayments/api/JSONBuilder.java
package com.braintreepayments.api;

import org.json.JSONException;
import org.json.JSONObject;

public abstract class JSONBuilder {

    protected JSONObject jsonBody;

    protected JSONBuilder(JSONObject json) {
        jsonBody = json;
    }

    public JSONBuilder() {
        jsonBody = new JSONObject();
    }

    public String build() {
        return jsonBody.toString();
    }

    public void put(Object value) {
        int stackIndex = 3;
        StackTraceElement[] stack = Thread.currentThread().getStackTrace();
        if (!stack[0].isNativeMethod()) {
            stackIndex--;
        }

        StackTraceElement current = stack[stackIndex];
        put(current.getMethodName(), value);
    }

    public void put(String key, Object value) {
        try {
            jsonBody.put(key, value);
        } catch (JSONException ignored) {}
    }
}

File: TestUtils/src/main/java/com/braintreepayments/api/testutils/JSONBuilder.kt
package com.braintreepayments.api.testutils

import org.json.JSONException
import org.json.JSONObject

abstract class JSONBuilder protected constructor(json: JSONObject = JSONObject()) {

    protected var jsonBody: JSONObject = json

    fun build(): String = jsonBody.toString()

    fun put(value: Any?) {
        var stackIndex = STACK_INDEX_CALLER
        val stack = Thread.currentThread().stackTrace
        if (!stack[0].isNativeMethod) {
            stackIndex--
        }
        val current = stack[stackIndex]
        put(current.methodName, value)
    }

    fun put(key: String, value: Any?) {
        try {
            jsonBody.put(key, value)
        } catch (_: JSONException) {
        }
    }

    companion object {
        private const val STACK_INDEX_CALLER = 3
    }
}
